presto-main/src/main/java/io/prestosql/execution/SqlTask.java

```diff
@@ -219,11 +230,18 @@ public TaskStatus getTaskStatus()
         }
     }
 
+    private void notifyStatusChanged()
+    {
+        taskStatusVersion.incrementAndGet();
+        taskStatusVersionChange.complete(null, taskNotificationExecutor);
+    }
+
     private TaskStatus createTaskStatus(TaskHolder taskHolder)
     {
-        // Always return a new TaskStatus with a larger version number;
-        // otherwise a client will not accept the update
-        long versionNumber = nextTaskStatusVersion.getAndIncrement();
+        // Obtain task status version before building actual TaskStatus object.
+        // This way any task updates won't be lost since all updates happen
+        // before version number is increased.
+        long versionNumber = taskStatusVersion.get();
```

Comment:
is it fine if `taskStateMachine.getState();` in the next line gives a newly updated state (let's say `versionNumber` is set to V which corresponds to a state S1 - but before the `getState` call executes in the next line, the state is changed to S2 with current versionNumber as V+1?

Reply:
Yes. That works. It might happen that client will get newer status (task state, DF version) with older version number. In such case client will make another request and get same status again. That isn't a problem. It is important that newest version of `TaskStatus` will be received eventually.
`versionNumber` isn't meant for consistency here, but rather for notifications.